### PR TITLE
X.P.OrgMode: Fix behaviour of getLast

### PR DESCRIPTION
**Pull request by slotThe**
### Description

It's funny to notice this so late; I guess non-deterministic tests do have their drawback :)

So far, while parsing strings like "<ptn><more-letters>", the `getLast`
function immediately stopped on strings of the form "<ptn><whitespace>".
This may be a bit confusing given the functions name.  Strings of
the—perhaps artificial—form

    "a +d f 1 +d f 2"

would be cut short and parsed as

    Deadline "a" (Time { date = Next Friday, tod = Just 01:00 })

instead of the more intuitive

    Deadline "a +d f 1" (Time { date = Next Friday, tod = Just 02:00 }).

This is readily fixed by applying the `go` parser as often as possible,
only returning the longest list, and then pruning eventual leftovers at
the end of the string.  Since we were already invoking `dropWhileEnd` to
trim whitespace before, the added `reverse`s should not impact
performance at all.

Fixes: https://github.com/xmonad/xmonad-contrib/issues/584

```
λ ⟩ pInput "a +d f 1 +d f 2"
Just (Deadline "a +d f 1" (Time {date = Next Friday, tod = Just 02:00}))
```

```
λ ⟩ quickCheck $ prop_decodeLinearity (Deadline "+d" (Time {date = XMonad.Prompt.OrgMode.Next Friday, tod = Nothing}))
+++ OK, passed 1 test.
```

```
λ ⟩ quickCheck $ prop_decodeLinearity (Deadline "+d f 1" (Time {date = XMonad.Prompt.OrgMode.Next Friday, tod = Nothing}))
+++ OK, passed 1 test.
```

### Checklist

  - [x] I've read [CONTRIBUTING.md](https://github.com/xmonad/xmonad/blob/master/CONTRIBUTING.md)

  - [x] I've considered how to best test these changes (property, unit,
        manually, ...) and concluded: I added some regression tests for the cases that were mentioned in #584

  - [n/a] I updated the `CHANGES.md` file
    Unreleased module, I don't think this is necessary.

  - [n/a] I updated the `XMonad.Doc.Extending` file (if appropriate)

### CODE REVIEW

File: XMonad/Prompt/OrgMode.hs [hunk 1]
```diff
@@ -435,3 +436,13 @@ pInt = read <$> munch1 isDigit
 -- parsing when the left-most parser succeeds.
 lchoice :: [ReadP a] -> ReadP a
 lchoice = foldl' (<++) empty
+
+-- | Like 'Text.ParserCombinators.ReadP', but only return the parse
```
Comment (liskin) [inline XMonad/Prompt/OrgMode.hs:440]:
```suggestion
-- | Like 'Text.ParserCombinators.ReadP.endBy1', but only return the parse
```

File: tests/OrgMode.hs [hunk 2]
```diff
@@ -22,6 +22,14 @@ spec = do
   prop "prop_encodeLinearity" prop_encodeLinearity
   prop "prop_decodeLinearity" prop_decodeLinearity
 
+  -- Checking for regressions
+  context "+d f +d f" $ do
+    it "encode" $ prop_encodeLinearity (OrgMsg "+d f +d f")
+    it "decode" $ prop_decodeLinearity (Deadline "+d" (Time {date = Next Friday, tod = Nothing}))
```
Comment (liskin) [inline tests/OrgMode.hs:28]:
```suggestion
    it "decode" $ prop_decodeLinearity (Deadline "+d f" (Time {date = Next Friday, tod = Nothing}))
```
Otherwise seems unrelated to the context.

(But then we might want to check the "+d" alone as well?)

**Review by liskin — APPROVED** [review]
I don't know enough about ReadP to confidently say this is bullet-proof, but it does look better than it was. :-)

**Review by psibi — APPROVED** [review]

File: tests/OrgMode.hs (hunk 2 above)
Comment (slotThe) [reply tests/OrgMode.hs:28]: Good point; I changed everything to `+d +d f` instead; we test for "someting else coming after the first `+d`" below anyways

**Comment by liskin** [comment]
I wanted to merge this today but then life interfered and I'll be afk most of tomorrow as well so please do feel free to `rebase --autosquash` and merge any time.
